### ISSUE TITLE
Run it in real time 

### ISSUE THREAD

**Issue by intvo** [harshilpatel312/KITTI-distance-estimation#2]
How can i run it in real time to estimate the distance from the car camera?

**Comment by harshilpatel312** [comment]
Good question, forgot to add that part in the project. I'll add it soon.

The way I would do it is:
1. Save StandardScalar's values from train.py (separate scalars for X and y) so that the real-time data can be scaled and passed to the network. (check how to save StandardScalar [here)](https://stackoverflow.com/a/41993653/8094534)
2. Load model and perform inverse scaling as done in inference.py.

It would look something like this (pseudocode):
```python
stream = cv2.VideoCapture('xyz')

# load scalars
scalar_x = joblib.load('scalar_x.p') 
scalar_y = joblib.load('scalar_y.p') 

# load model and weights
loaded_model = model_from_json( loaded_model_json )
loaded_model.load_weights("generated_files/{}.h5".format(WEIGHTS))
loaded_model.compile(loss='mean_squared_error', optimizer='adam')

# perform distance estimation
while True:
    _, image = stream.read()
    bboxs = perform_object_detection(image)
    scaled_bboxs = scalar_x.transform(bboxs)
    y_pred = loaded_model.predict(scaled_bboxs)
    range, bearing = scalar.inverse_transform(y_pred)
```

**Comment by xLobito** [comment]
> 
> 
> Good question, forgot to add that part in the project. I'll add it soon (or if you could make a pull request, I'd be happy to integrate it).
> 
Well, I tried to replicate the pseudocode and when I try to inverse transform the predicted value from a coordinates (bounding box of a car) the value is negative (values like -22 or -270 in other cases). 

So, I'm wondering if there's something else behind? Maybe I did something wrong in the process. 

I use an image and set the bounding box accordingly as an array to be passed to transform and the use the scaler_y to inverse it. 
![imagen](https://user-images.githubusercontent.com/16466974/70683635-d8b01a00-1c81-11ea-8737-974a8db6a801.png)

Thanks.

**Comment by OctaM** [comment]
If the images you are trying to predict have a different shape from them ones in KITTI, you will get negative/strange results because the aspect ratio is different. 

You should resize your images so they match the shape from the dataset, which is (1242, 375) and you will get results in >= 0, but they may not be as accurate as you need.

**Comment by danammeansbear** [comment]
Was real time ever added? Thank you for including the pseudo code to get me started!!

**Comment by danammeansbear** [comment]
`# -*- coding: utf-8 -*-
"""
Created on Sun Jan 10 12:16:46 2021

@author: fahre
"""

import pandas as pd
import argparse

import cv2
from keras.models import model_from_json
from sklearn.preprocessing import StandardScaler

argparser = argparse.ArgumentParser(description='Get predictions of test set')
argparser.add_argument('-m', '--modelname',
                       help='model name (.json)')
argparser.add_argument('-w', '--weights',
                       help='weights filename (.h5)')

args = argparser.parse_args()
stream = cv2.VideoCapture('xyz')
# parse arguments
MODEL = args.modelname
WEIGHTS = args.weights

def main():
    # get data
    df_test = pd.read_csv('data/test.csv')
    X_test = df_test[['xmin', 'ymin', 'xmax', 'ymax']].values
    y_test = df_test[['zloc']].values

    # standardized data
    scalar = StandardScaler()
    X_test = scalar.fit_transform(X_test)
    y_test = scalar.fit_transform(y_test)

    # load json and create model
    json_file = open('generated_files/{}.json'.format(MODEL), 'r')
    loaded_model_json = json_file.read()
    json_file.close()
    loaded_model = model_from_json( loaded_model_json )

    # load weights into new model
    loaded_model.load_weights("generated_files/{}.h5".format(WEIGHTS))
    print("Loaded model from disk")

    # evaluate loaded model on test data
    loaded_model.compile(loss='mean_squared_error', optimizer='adam')
    y_pred = loaded_model.predict(X_test)
    while True:
        _, image = stream.read()
        bboxs = perform_object_detection(image)
        scaled_bboxs = X_test.transform(bboxs)
        y_pred = loaded_model.predict(scaled_bboxs)
        range, bearing = y_test.inverse_transform(y_pred)

    # scale up predictions to original values
    y_pred = scalar.inverse_transform(y_pred)
    y_test = scalar.inverse_transform(y_test)

    # save predictions
    df_result = df_test
    df_result['zloc_pred'] = -100000

    for idx, row in df_result.iterrows():
        df_result.at[idx, 'zloc_pred'] = y_pred[idx]

    df_result.to_csv('data/predictions.csv', index=False)
    # perform distance estimation
    

if __name__ == '__main__':
    main()
`

I have the following attempt at the code, I just get a error when i use the line         bboxs = perform_object_detection(image)

**Comment by harshilpatel312** [comment]
I'll take a look in the next week if I get a chance..no promises though sorry..

If someone wants to submit a pull request, I'll be happy to look into it.
